include/fiction/algorithms/simulation/sidb/critical_temperature.hpp

```diff
@@ -412,94 +323,158 @@ class critical_temperature_impl
+ * @tparam TT The type of the truth table specifying the gate behavior.
  * @param lyt The layout to simulate.
+ * @param spec Expected Boolean function of the layout given as a multi-output truth table.
  * @param params Simulation and physical parameters.
  * @param pst Statistics.
  */
-template <typename Lyt>
-bool critical_temperature(const Lyt& lyt, const critical_temperature_params& params = {},
-                          critical_temperature_stats<Lyt>* pst = nullptr)
+template <typename Lyt, typename TT>
+bool critical_temperature_gate_based(const Lyt& lyt, const std::vector<TT>& spec,
+                                     const critical_temperature_params& params = {},
+                                     critical_temperature_stats<Lyt>*   pst    = nullptr)
 {
     static_assert(is_cell_level_layout_v<Lyt>, "Lyt is not a cell-level layout");
     static_assert(has_sidb_technology_v<Lyt>, "Lyt is not an SiDB layout");
     static_assert(has_siqad_coord_v<Lyt>, "Lyt is not based on SiQAD coordinates");
 
+    assert(!spec.empty());
+    // all elements in tts must have the same number of variables
+    assert(std::adjacent_find(spec.begin(), spec.end(),
+                              [](const auto& a, const auto& b) { return a.num_vars() != b.num_vars(); }) == spec.end());
+
     critical_temperature_stats<Lyt> st{};
 
-    detail::critical_temperature_impl<Lyt> p{lyt, params, st};
+    detail::critical_temperature_impl<Lyt, TT> p{lyt, spec, params, st};
 
-    bool result = false;
+    const auto result = p.gate_based_simulation();
 
-    if (params.temperature_mode == critical_temperature_params::critical_temperature_mode::GATE_BASED_SIMULATION)
-    {
-        result = p.gate_based_simulation();
-    }
-    else
+    if (pst)
     {
-        result = p.non_gate_based_simulation();
+        *pst = st;
     }
 
+    return result;
+}
+
+/**
+ * For *Non-gate-based Critical Temperature* simulation, the Critical Temperature is defined as follows: The temperature
+ * at which the excited charge distributions are populated by more than \f$ 1 - \eta \f$, where \f$ \eta \in [0,1] \f$
+ * is the confidence level for the presence of a working gate.
+ *
+ * @tparam Lyt SiDB cell-level layout type.
+ * @param lyt The layout to simulate.
+ * @param params Simulation and physical parameters.
+ * @param pst Statistics.
+ */
+template <typename Lyt>
+bool critical_temperature_non_gate_based(const Lyt& lyt, const critical_temperature_params& params = {},
+                                         critical_temperature_stats<Lyt>* pst = nullptr)
+{
+    static_assert(is_cell_level_layout_v<Lyt>, "Lyt is not a cell-level layout");
+    static_assert(has_sidb_technology_v<Lyt>, "Lyt is not an SiDB layout");
+    static_assert(has_siqad_coord_v<Lyt>, "Lyt is not based on SiQAD coordinates");
+
+    critical_temperature_stats<Lyt> st{};
+
+    detail::critical_temperature_impl<Lyt, tt> p{lyt, std::vector<tt>{tt{}}, params, st};
```

Comment:
This is a bit cleaner than before but still involves passing an empty truth table to an object that doesn't need one at all. I would refactor `critical_temperature_impl` to no longer have that truth table attribute and instead pass the truth table to `gate_based_simulation`.

Please see `operational_domain.hpp` again for an example of how we deal with different functions with varying parameters but the same underlying `_impl` class.

Reply:
@marcelwa ah, true. Now I can see it in ``operational_domain.hpp``. That's great. Thank you very much!